### PR TITLE
Add Redis Storage Integration Tests with Testcontainers

### LINKED ISSUE #3629: Auth Server: Add Redis Storage Integration Tests with Testcontainers (Phase 2)
## Description

Add integration tests for the Redis storage implementation that run against real Redis Sentinel instances using testcontainers. This ensures the `RedisStorage` implementation works correctly with actual Redis behavior, including Sentinel failover, ACL authentication, and proper TTL handling that cannot be fully tested with the miniredis mock.

## Context

TASK-001 implemented the `RedisStorage` struct with comprehensive unit tests using miniredis. While miniredis is excellent for fast unit testing, it does not fully replicate Redis Sentinel behavior, ACL user authentication, or real network conditions. Integration tests against real Redis are essential to validate the implementation works correctly in production-like environments.

**Parent Issue**: https://github.com/stacklok/stacklok-epics/issues/197
**Dependencies**: #3628 (Redis Storage: Core Implementation - Phase 1)
**Blocks**: TASK-003 (Documentation & Operator CRD Updates - Phases 3+4)

## Acceptance Criteria

- [ ] Integration tests run against real Redis Sentinel setup via testcontainers
- [ ] Tests cover all `Storage` interface methods with real Redis
- [ ] Tests verify Sentinel failover behavior (primary switchover)
- [ ] Tests verify ACL user authentication
- [ ] Tests verify TTL-based expiration with real Redis
- [ ] Tests are tagged appropriately for CI (e.g., `//go:build integration`)
- [ ] `task test-all` passes (or tests are skipped when testcontainers/Docker unavailable)
- [ ] Code follows existing integration test patterns in `test/integration/authserver/`

### Recommended Implementation

Create a testcontainers-based setup that spins up a Redis Sentinel cluster (1 primary + 2 replicas + 3 sentinels) and runs the full `Storage` interface test suite against it. Follow the existing integration test patterns in `test/integration/authserver/` and `pkg/authserver/storage/memory_test.go`.

**Key considerations:**
1. Use testcontainers-go to manage Redis containers
2. Configure Redis Sentinel with ACL user authentication
3. Reuse test helpers and patterns from `memory_test.go` (mock types, table-driven tests)
4. Add build tags to separate from unit tests
5. Handle Docker availability gracefully (skip tests when unavailable)

### Patterns & Frameworks

- **testcontainers-go**: Container management for integration tests (`github.com/testcontainers/testcontainers-go`)
- **Redis Sentinel module**: Use testcontainers Redis module or custom container setup for Sentinel
- **Build tags**: Use `//go:build integration` to separate from unit tests
- **Table-driven tests**: Follow patterns from `memory_test.go` for comprehensive coverage
- **Test fixtures**: Follow patterns from `test/integration/authserver/helpers/` for reusable test setup

### Code Pointers

- `pkg/authserver/storage/memory_test.go` - Test patterns, mock types (`mockSession`, `mockClient`, `mockRequester`), and table-driven test structure to follow
- `pkg/authserver/storage/types.go` - `Storage` interface definition (all methods to test)
- `pkg/authserver/storage/redis.go` - Redis storage implementation to test (created in TASK-001)
- `test/integration/authserver/helpers/authserver.go` - Integration test helper patterns to follow
- `test/integration/authserver/authserver_integration_test.go` - Example integration test structure
- `docs/proposals/THV-0035-auth-server-redis-storage.md` - RFC with testing strategy

### Component Interfaces

**Test Container Setup**

```go
// pkg/authserver/storage/redis_integration_test.go

//go:build integration

package storage_test

import (
    "context"
    "testing"
    "time"

    "github.com/testcontainers/testcontainers-go"
    // ... other imports
)

// RedisSentinelCluster represents a test Redis Sentinel cluster.
type RedisSentinelCluster struct {
    PrimaryContainer   testcontainers.Container
    Replica1Container  testcontainers.Container
    Replica2Container  testcontainers.Container
    Sentinel1Container testcontainers.Container
    Sentinel2Container testcontainers.Container
    Sentinel3Container testcontainers.Container

    SentinelAddrs []string
    MasterName    string
    Username      string
    Password      string
}

// NewRedisSentinelCluster creates a Redis Sentinel cluster for testing.
// Returns a cluster with 1 primary + 2 replicas + 3 sentinels.
func NewRedisSentinelCluster(ctx context.Context, t *testing.T) *RedisSentinelCluster

// Close terminates all containers.
func (c *RedisSentinelCluster) Close(ctx context.Context) error

// TriggerFailover forces a Sentinel failover for testing.
func (c *RedisSentinelCluster) TriggerFailover(ctx context.Context) error
```

**Storage Test Helper**

```go
// withRedisStorage creates a RedisStorage for testing and handles cleanup.
func withRedisStorage(t *testing.T, cluster *RedisSentinelCluster, fn func(context.Context, *RedisStorage)) {
    t.Helper()
    ctx := context.Background()

    cfg := RedisConfig{
        SentinelConfig: &SentinelConfig{
            MasterName:    cluster.MasterName,
            SentinelAddrs: cluster.SentinelAddrs,
            DB:            0,
        },
        ACLUserConfig: &ACLUserConfig{
            Username: cluster.Username,
            Password: cluster.Password,
        },
        KeyPrefix: DeriveKeyPrefix("test-ns", "test-server"),
    }

    storage, err := NewRedisStorage(ctx, cfg)
    require.NoError(t, err)
    defer storage.Close()

    fn(ctx, storage)
}
```

### Testing Strategy

**Integration Tests** (using testcontainers)

Storage Interface Methods:
- [ ] `RegisterClient` / `GetClient` - Client registration and retrieval
- [ ] `ClientAssertionJWTValid` / `SetClientAssertionJWT` - JWT assertion tracking
- [ ] `CreateAuthorizeCodeSession` / `GetAuthorizeCodeSession` / `InvalidateAuthorizeCodeSession` - Auth code flow
- [ ] `CreateAccessTokenSession` / `GetAccessTokenSession` / `DeleteAccessTokenSession` - Access tokens
- [ ] `CreateRefreshTokenSession` / `GetRefreshTokenSession` / `DeleteRefreshTokenSession` - Refresh tokens
- [ ] `RotateRefreshToken` - Token rotation
- [ ] `RevokeAccessToken` / `RevokeRefreshToken` / `RevokeRefreshTokenMaybeGracePeriod` - Token revocation (RFC 7009)
- [ ] `CreatePKCERequestSession` / `GetPKCERequestSession` / `DeletePKCERequestSession` - PKCE flow
- [ ] `StoreUpstreamTokens` / `GetUpstreamTokens` / `DeleteUpstreamTokens` - Upstream IDP tokens
- [ ] `StorePendingAuthorization` / `LoadPendingAuthorization` / `DeletePendingAuthorization` - Pending auth
- [ ] `CreateUser` / `GetUser` / `DeleteUser` - User management
- [ ] `CreateProviderIdentity` / `GetProviderIdentity` / `UpdateProviderIdentityLastUsed` / `GetUserProviderIdentities` - Provider identities

Sentinel-Specific Tests:
- [ ] Test connection via Sentinel addresses (not direct Redis)
- [ ] Test automatic reconnection after primary failover
- [ ] Test operation continuity during failover
- [ ] Test quorum-based failover detection

ACL Authentication Tests:
- [ ] Test successful authentication with valid ACL credentials
- [ ] Test authentication failure with invalid username
- [ ] Test authentication failure with invalid password
- [ ] Test key pattern restrictions (operations on `thv:auth:*` succeed, others fail)

TTL and Expiration Tests:
- [ ] Test token expiration via Redis TTL (not cleanup goroutine)
- [ ] Test that expired tokens return appropriate errors
- [ ] Verify TTL values match session expiration times

Concurrent Access Tests:
- [ ] Test concurrent writes to different keys
- [ ] Test concurrent reads and writes
- [ ] Test concurrent client registration and lookup

**Edge Cases**

- [ ] Connection timeout handling
- [ ] Network partition recovery
- [ ] Redis restart during operation
- [ ] Large payload serialization/deserialization
- [ ] Unicode characters in user IDs and provider subjects

## Out of Scope

- E2E tests for the full auth server with Redis (covered in future work)
- Operator/CRD integration tests (covered in TASK-003 and Phase 5)
- Performance benchmarks (future enhancement)
- TLS/mTLS testing for Redis connections (future enhancement per RFC)
- Multi-tenant isolation testing across different key prefixes (covered by unit tests)

## References

- Epic: https://github.com/stacklok/stacklok-epics/issues/197
- Upstream Dependency: #3628 (Redis Storage: Core Implementation)
- RFC Design Document: `docs/proposals/THV-0035-auth-server-redis-storage.md`
- Storage Interface: `pkg/authserver/storage/types.go`
- Memory Storage Tests: `pkg/authserver/storage/memory_test.go`
- Auth Server Integration Tests: `test/integration/authserver/`
- testcontainers-go: https://golang.testcontainers.org/
- Redis Sentinel Documentation: https://redis.io/docs/management/sentinel/

### PR DESCRIPTION
**Pull request by tgrunnagle**
Depends on #3639 
Closes #3629

## Summary

Adds comprehensive integration tests for the Redis Sentinel storage backend using testcontainers-go. These tests validate the `RedisStorage` implementation against a real Redis Sentinel cluster (1 primary + 2 replicas + 3 sentinels), covering behavior that cannot be verified with the miniredis mock used in unit tests—including Sentinel failover, ACL authentication, real TTL expiration, and concurrent access under actual network conditions.

### Integration Test Suite (`pkg/authserver/storage/redis_integration_test.go`)
- Added testcontainers-based Redis Sentinel cluster setup (1 primary, 2 replicas, 3 sentinels) with Docker network isolation
- Configured ACL user authentication (`~thv:*` key pattern restriction) on all Redis nodes
- Implemented custom `Dialer` address translation to work around Docker-internal IPs being unreachable from macOS hosts
- Used `TestMain` for shared cluster lifecycle with graceful skip when Docker is unavailable

### Storage Interface Coverage
- Full CRUD tests for all `Storage` interface methods: clients, authorization codes, access tokens, refresh tokens, PKCE, upstream tokens, pending authorizations, users, and provider identities
- Token rotation and revocation tests (including grace period revocation)
- Session round-trip test verifying JWT claims survive serialization through real Redis
- Cascade deletion test (deleting a user removes associated identities and upstream tokens)

### Sentinel-Specific Tests
- Sentinel connection verification (write/read through Sentinel-routed connection)
- Failover test: writes data, waits for replication (`WAIT 1 5000`), triggers failover, verifies data survives and new writes succeed on the promoted primary

### ACL Authentication Tests
- Valid credentials succeed for `thv:*` key operations
- Invalid username and invalid password are rejected
- Key pattern restriction enforced (writes to non-`thv:*` keys are denied)

### TTL and Expiration Tests
- Access tokens expire via real Redis TTL (not cleanup goroutine)
- JTI entries expire and become valid again after TTL
- Redis TTL values match session expiration times

### Concurrent Access Tests
- 50 concurrent writes to different keys with verification
- Mixed concurrent reads and writes
- Concurrent client registration and lookup

### Edge Cases
- Unicode characters in user IDs and provider subjects (CJK, accented, emoji)

## Implementation Details

- Tests use `//go:build integration` tag, excluded from `task test` by default
- Each subtest gets an isolated key prefix via `DeriveKeyPrefix("inttest", sanitizedTestName)` to prevent cross-test interference
- All subtests run in parallel via the `withIntegrationStorage` helper
- The failover test is intentionally non-parallel since it modifies cluster topology

## Testing

- 18 test functions covering all `Storage` interface methods plus Sentinel, ACL, TTL, concurrency, and Unicode edge cases
- Run with: `go test -tags integration -v -timeout 5m ./pkg/authserver/storage/`
- Tests skip gracefully when Docker is unavailable
- `testcontainers-go` dependency added to `go.mod`

## Large PR Justification

- Adds integration test suite for a single interface

### CODE REVIEW

File: pkg/authserver/storage/redis_integration_test.go [hunk 1]
```diff
@@ -0,0 +1,1213 @@
...
+
+func TestMain(m *testing.M) {
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Minute)
+	var err error
+	testCluster, err = newRedisSentinelCluster(ctx)
+	cancel()
+
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Warning: could not set up Redis Sentinel cluster: %v\n", err)
+		fmt.Fprintf(os.Stderr, "Integration tests will be skipped; unit tests will still run.\n")
```
Comment (jhrozek) [inline pkg/authserver/storage/redis_integration_test.go]: 2 comments in one:
 - the tests have a `//go:build integration` build constraint so they don't actually get included in `task test`. Should there be a separate task target and/or github workflow so the integration tests are executed as part of the CI?
 - if yes, should the warning turn into a hard error?

File: pkg/authserver/storage/redis_integration_test.go [hunk 2]
```diff
@@ -0,0 +1,1213 @@
...
+		t.Skip("Redis Sentinel cluster not available")
+	}
+	t.Parallel()
+
+	client := testCluster.newTestClient()
+	prefix := DeriveKeyPrefix("inttest", sanitizeTestName(t.Name()))
+	storage := NewRedisStorageWithClient(client, prefix)
+	t.Cleanup(func() { _ = storage.Close() })
+
+	fn(context.Background(), storage)
```
Comment (jhrozek) [inline pkg/authserver/storage/redis_integration_test.go]: should we wrap context.Background in a WithTimeout context?

File: pkg/authserver/storage/redis_integration_test.go [hunk 3]
```diff
@@ -0,0 +1,1213 @@
...
+	req := testcontainers.ContainerRequest{
+		Image:        testRedisImage,
+		ExposedPorts: []string{"26379/tcp"},
+		Networks:     []string{networkName},
+		Cmd:          []string{"redis-sentinel", "/data/sentinel.conf"},
+		Files: []testcontainers.ContainerFile{
+			{
+				Reader:            strings.NewReader(config),
+				ContainerFilePath: "/data/sentinel.conf",
+				FileMode:          0o666,
```
Comment (jhrozek) [inline pkg/authserver/storage/redis_integration_test.go]: It's probably a non-issue in a test shortlived setup but `rw-rw-rw-` will always trigger my spider-sense :-) can we use 0o644?

(sorry for a nit)

**Review by jhrozek — APPROVED** [review]
I think it's fine merging the test as-is but it wasn't clear to me how the tests are ran..

File: pkg/authserver/storage/redis_integration_test.go (hunk 1 above)
Comment (tgrunnagle) [reply pkg/authserver/storage/redis_integration_test.go]: I will add a task target, I think gh workflow might warrant it's own PR
